[TEST-FIX] ConditionTest.testLockConditionSignalAllShutDownKeyOwner

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-problematicTest/com.hazelcast$hazelcast/223/testReport/junit/com.hazelcast.concurrent.lock/ConditionTest/testLockConditionSignalAllShutDownKeyOwner/
```
java.lang.AssertionError: expected:<50> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at com.hazelcast.concurrent.lock.ConditionTest.testLockConditionSignalAllShutDownKeyOwner(ConditionTest.java:426)
```

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/ConditionKey.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/ConditionKey.java
@@ -24,16 +24,24 @@ public final class ConditionKey implements WaitNotifyKey {
     private final String name;
     private final Data key;
     private final String conditionId;
+    private final long threadId;
+    private final String uuid;
 
-    public ConditionKey(String name, Data key, String conditionId) {
+    public ConditionKey(String name, Data key, String conditionId, String uuid, long threadId) {
         this.name = name;
         this.key = key;
         this.conditionId = conditionId;
+        this.uuid = uuid;
+        this.threadId = threadId;
     }
 
     @Override
     public String getServiceName() {
         return LockServiceImpl.SERVICE_NAME;
+    }
+
+    public String getUuid() {
+        return uuid;
     }
 
     @Override
@@ -49,31 +57,53 @@ public final class ConditionKey implements WaitNotifyKey {
         return conditionId;
     }
 
+    public long getThreadId() {
+        return threadId;
+    }
+
     @Override
+    @SuppressWarnings("checkstyle:npathcomplexity")
     public boolean equals(Object o) {
         if (this == o) {
             return true;
         }
-        if (o == null || getClass() != o.getClass()) {
+        if (!(o instanceof ConditionKey)) {
             return false;
         }
-
         ConditionKey that = (ConditionKey) o;
-
-        if (key != null ? !key.equals(that.key) : that.key != null) {
+        if (threadId != that.threadId) {
             return false;
         }
-        if (conditionId != null ? !conditionId.equals(that.conditionId) : that.conditionId != null) {
+        if (!name.equals(that.name)) {
             return false;
         }
+        if (!key.equals(that.key)) {
+            return false;
+        }
+        if (!uuid.equals(that.uuid)) {
+            return false;
+        }
+        return conditionId.equals(that.conditionId);
 
-        return true;
     }
 
     @Override
     public int hashCode() {
-        int result = key != null ? key.hashCode() : 0;
-        result = 31 * result + (conditionId != null ? conditionId.hashCode() : 0);
+        int result = name.hashCode();
+        result = 31 * result + key.hashCode();
+        result = 31 * result + conditionId.hashCode();
+        result = 31 * result + (int) (threadId ^ (threadId >>> 32));
+        result = 31 * result + uuid.hashCode();
         return result;
+    }
+
+    @Override
+    public String toString() {
+        return "ConditionKey{"
+                + "name='" + name + '\''
+                + ", key=" + key
+                + ", conditionId='" + conditionId + '\''
+                + ", threadId=" + threadId
+                + '}';
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockResourceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockResourceImpl.java
@@ -27,10 +27,12 @@ import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 
 final class LockResourceImpl implements DataSerializable, LockResource {
 
@@ -43,8 +45,8 @@ final class LockResourceImpl implements DataSerializable, LockResource {
     private long acquireTime = -1L;
     private boolean transactional;
     private boolean blockReads;
-    private Map<String, ConditionInfo> conditions;
-    private List<ConditionKey> signalKeys;
+    private Map<String, WaitersInfo> waiters;
+    private Set<ConditionKey> conditionKeys;
     private List<AwaitOperation> expiredAwaitOps;
     private LockStoreImpl lockStore;
 
@@ -160,71 +162,74 @@ final class LockResourceImpl implements DataSerializable, LockResource {
         return lockCount == 0 || getThreadId() == threadId && getOwner().equals(caller);
     }
 
-    boolean addAwait(String conditionId, String caller, long threadId) {
-        if (conditions == null) {
-            conditions = new HashMap<String, ConditionInfo>(2);
+    void addAwait(String conditionId, String caller, long threadId) {
+        if (waiters == null) {
+            waiters = new HashMap<String, WaitersInfo>(2);
         }
 
-        ConditionInfo condition = conditions.get(conditionId);
+        WaitersInfo condition = waiters.get(conditionId);
         if (condition == null) {
-            condition = new ConditionInfo(conditionId);
-            conditions.put(conditionId, condition);
+            condition = new WaitersInfo(conditionId);
+            waiters.put(conditionId, condition);
         }
-        return condition.addWaiter(caller, threadId);
+        condition.addWaiter(caller, threadId);
     }
 
-    boolean removeAwait(String conditionId, String caller, long threadId) {
-        if (conditions == null) {
-            return false;
+    void removeAwait(String conditionId, String caller, long threadId) {
+        if (waiters == null) {
+            return;
         }
 
-        ConditionInfo condition = conditions.get(conditionId);
+        WaitersInfo condition = waiters.get(conditionId);
         if (condition == null) {
-            return false;
+            return;
         }
 
-        boolean ok = condition.removeWaiter(caller, threadId);
-        if (condition.getAwaitCount() == 0) {
-            conditions.remove(conditionId);
+        condition.removeWaiter(caller, threadId);
+        if (!condition.hasWaiter()) {
+            waiters.remove(conditionId);
         }
-        return ok;
     }
 
-    boolean startAwaiting(String conditionId, String caller, long threadId) {
-        if (conditions == null) {
-            return false;
+    public void signal(String conditionId, int maxSignalCount) {
+        if (waiters == null) {
+            return;
         }
 
-        ConditionInfo condition = conditions.get(conditionId);
+        Set<WaitersInfo.ConditionWaiter> waiters;
+        WaitersInfo condition = this.waiters.get(conditionId);
         if (condition == null) {
-            return false;
-        }
-
-        return condition.startWaiter(caller, threadId);
-    }
-
-    int getAwaitCount(String conditionId) {
-        if (conditions == null) {
-            return 0;
-        }
-
-        ConditionInfo condition = conditions.get(conditionId);
-        if (condition == null) {
-            return 0;
+            return;
         } else {
-            return condition.getAwaitCount();
+            waiters = condition.getWaiters();
         }
+        if (waiters == null) {
+            return;
+        }
+        String objectName = lockStore.getNamespace().getObjectName();
+        Iterator<WaitersInfo.ConditionWaiter> iterator = waiters.iterator();
+        for (int i = 0; iterator.hasNext() && i < maxSignalCount; i++) {
+            WaitersInfo.ConditionWaiter waiter = iterator.next();
+            ConditionKey signalKey = new ConditionKey(objectName,
+                    key, conditionId, waiter.getCaller(), waiter.getThreadId());
+            registerSignalKey(signalKey);
+            iterator.remove();
+        }
+        if (!condition.hasWaiter()) {
+            this.waiters.remove(conditionId);
+        }
+
     }
 
-    void registerSignalKey(ConditionKey conditionKey) {
-        if (signalKeys == null) {
-            signalKeys = new LinkedList<ConditionKey>();
+    private void registerSignalKey(ConditionKey conditionKey) {
+        if (conditionKeys == null) {
+            conditionKeys = new HashSet<ConditionKey>();
         }
-        signalKeys.add(conditionKey);
+        conditionKeys.add(conditionKey);
     }
 
     ConditionKey getSignalKey() {
-        List<ConditionKey> keys = signalKeys;
+        Set<ConditionKey> keys = conditionKeys;
         if (isNullOrEmpty(keys)) {
             return null;
         }
@@ -233,9 +238,16 @@ final class LockResourceImpl implements DataSerializable, LockResource {
     }
 
     void removeSignalKey(ConditionKey conditionKey) {
-        if (signalKeys != null) {
-            signalKeys.remove(conditionKey);
+        if (conditionKeys != null) {
+            conditionKeys.remove(conditionKey);
         }
+    }
+
+    boolean hasSignalKey(ConditionKey conditionKey) {
+        if (conditionKeys == null) {
+            return false;
+        }
+        return conditionKeys.contains(conditionKey);
     }
 
     void registerExpiredAwaitOp(AwaitOperation awaitResponse) {
@@ -276,8 +288,9 @@ final class LockResourceImpl implements DataSerializable, LockResource {
 
     boolean isRemovable() {
         return !isLocked()
-                && isNullOrEmpty(conditions)
-                && isNullOrEmpty(expiredAwaitOps);
+                && isNullOrEmpty(waiters)
+                && isNullOrEmpty(expiredAwaitOps)
+                && isNullOrEmpty(conditionKeys);
     }
 
     @Override
@@ -354,16 +367,18 @@ final class LockResourceImpl implements DataSerializable, LockResource {
         int conditionCount = getConditionCount();
         out.writeInt(conditionCount);
         if (conditionCount > 0) {
-            for (ConditionInfo condition : conditions.values()) {
+            for (WaitersInfo condition : waiters.values()) {
                 condition.writeData(out);
             }
         }
         int signalCount = getSignalCount();
         out.writeInt(signalCount);
         if (signalCount > 0) {
-            for (ConditionKey signalKey : signalKeys) {
+            for (ConditionKey signalKey : conditionKeys) {
                 out.writeUTF(signalKey.getObjectName());
                 out.writeUTF(signalKey.getConditionId());
+                out.writeUTF(signalKey.getUuid());
+                out.writeLong(signalKey.getThreadId());
             }
         }
         int expiredAwaitOpsCount = getExpiredAwaitsOpsCount();
@@ -380,11 +395,11 @@ final class LockResourceImpl implements DataSerializable, LockResource {
     }
 
     private int getSignalCount() {
-        return signalKeys == null ? 0 : signalKeys.size();
+        return conditionKeys == null ? 0 : conditionKeys.size();
     }
 
     private int getConditionCount() {
-        return conditions == null ? 0 : conditions.size();
+        return waiters == null ? 0 : waiters.size();
     }
 
     @Override
@@ -401,19 +416,19 @@ final class LockResourceImpl implements DataSerializable, LockResource {
 
         int len = in.readInt();
         if (len > 0) {
-            conditions = new HashMap<String, ConditionInfo>(len);
+            waiters = new HashMap<String, WaitersInfo>(len);
             for (int i = 0; i < len; i++) {
-                ConditionInfo condition = new ConditionInfo();
+                WaitersInfo condition = new WaitersInfo();
                 condition.readData(in);
-                conditions.put(condition.getConditionId(), condition);
+                waiters.put(condition.getConditionId(), condition);
             }
         }
 
         len = in.readInt();
         if (len > 0) {
-            signalKeys = new ArrayList<ConditionKey>(len);
+            conditionKeys = new HashSet<ConditionKey>(len);
             for (int i = 0; i < len; i++) {
-                signalKeys.add(new ConditionKey(in.readUTF(), key, in.readUTF()));
+                conditionKeys.add(new ConditionKey(in.readUTF(), key, in.readUTF(), in.readUTF(), in.readLong()));
             }
         }
 
@@ -470,5 +485,28 @@ final class LockResourceImpl implements DataSerializable, LockResource {
 
     private static boolean isNullOrEmpty(Map m) {
         return m == null || m.isEmpty();
+    }
+
+    void cleanWaitersAndSignalsFor(String uuid) {
+        if (conditionKeys != null) {
+            Iterator<ConditionKey> iter = conditionKeys.iterator();
+            while (iter.hasNext()) {
+                ConditionKey conditionKey = iter.next();
+                if (conditionKey.getUuid().equals(uuid)) {
+                    iter.remove();
+                }
+            }
+        }
+        if (waiters != null) {
+            for (WaitersInfo waitersInfo : waiters.values()) {
+                Iterator<WaitersInfo.ConditionWaiter> iter = waitersInfo.getWaiters().iterator();
+                while (iter.hasNext()) {
+                    WaitersInfo.ConditionWaiter waiter = iter.next();
+                    if (waiter.getCaller().equals(uuid)) {
+                        iter.remove();
+                    }
+                }
+            }
+        }
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockServiceImpl.java
@@ -179,7 +179,7 @@ public final class LockServiceImpl implements LockService, ManagedService, Remot
                 @Override
                 public void run() {
                     for (LockStoreImpl lockStore : container.getLockStores()) {
-                        releaseLock(operationService, uuid, container.getPartitionId(), lockStore);
+                        cleanUpLock(operationService, uuid, container.getPartitionId(), lockStore);
                     }
                 }
 
@@ -192,18 +192,19 @@ public final class LockServiceImpl implements LockService, ManagedService, Remot
         }
     }
 
-    private void releaseLock(OperationService operationService, String uuid, int partitionId, LockStoreImpl lockStore) {
+    private void cleanUpLock(OperationService operationService, String uuid, int partitionId, LockStoreImpl lockStore) {
         Collection<LockResource> locks = lockStore.getLocks();
         for (LockResource lock : locks) {
             Data key = lock.getKey();
             if (uuid.equals(lock.getOwner()) && !lock.isTransactional()) {
-                UnlockOperation op = createUnlockOperation(partitionId, lockStore.getNamespace(), key, uuid);
+                UnlockOperation op = createLockCleanupOperation(partitionId, lockStore.getNamespace(), key, uuid);
                 operationService.runOperationOnCallingThread(op);
             }
+            lockStore.cleanWaitersAndSignalsFor(key, uuid);
         }
     }
 
-    private UnlockOperation createUnlockOperation(int partitionId, ObjectNamespace namespace, Data key, String uuid) {
+    private UnlockOperation createLockCleanupOperation(int partitionId, ObjectNamespace namespace, Data key, String uuid) {
         UnlockOperation op = new LocalLockCleanupOperation(namespace, key, uuid);
         op.setAsyncBackup(true);
         op.setNodeEngine(nodeEngine);

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockStoreImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/LockStoreImpl.java
@@ -22,6 +22,7 @@ import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.nio.serialization.DataSerializable;
 import com.hazelcast.spi.ObjectNamespace;
+import com.hazelcast.spi.WaitNotifyKey;
 import com.hazelcast.util.ConcurrencyUtil;
 import com.hazelcast.util.ConstructorFunction;
 import com.hazelcast.util.scheduler.EntryTaskScheduler;
@@ -192,6 +193,11 @@ public final class LockStoreImpl implements DataSerializable, LockStore {
         }
     }
 
+    void cleanWaitersAndSignalsFor(Data key, String uuid) {
+        LockResourceImpl lockResource = locks.get(key);
+        lockResource.cleanWaitersAndSignalsFor(uuid);
+    }
+
     public int getVersion(Data key) {
         LockResourceImpl lock = locks.get(key);
         if (lock != null) {
@@ -258,32 +264,30 @@ public final class LockStoreImpl implements DataSerializable, LockStore {
         return backupCount + asyncBackupCount;
     }
 
-    public boolean addAwait(Data key, String conditionId, String caller, long threadId) {
+    public void addAwait(Data key, String conditionId, String caller, long threadId) {
         LockResourceImpl lock = getLock(key);
-        return lock.addAwait(conditionId, caller, threadId);
+        lock.addAwait(conditionId, caller, threadId);
     }
 
-    public boolean removeAwait(Data key, String conditionId, String caller, long threadId) {
+    public void removeAwait(Data key, String conditionId, String caller, long threadId) {
         LockResourceImpl lock = getLock(key);
-        return lock.removeAwait(conditionId, caller, threadId);
+        lock.removeAwait(conditionId, caller, threadId);
     }
 
-    public boolean startAwaiting(Data key, String conditionId, String caller, long threadId) {
+    public void signal(Data key, String conditionId, int maxSignalCount) {
         LockResourceImpl lock = getLock(key);
-        return lock.startAwaiting(conditionId, caller, threadId);
+        lock.signal(conditionId, maxSignalCount);
     }
 
-    public int getAwaitCount(Data key, String conditionId) {
-        LockResourceImpl lock = getLock(key);
-        return lock.getAwaitCount(conditionId);
+    public WaitNotifyKey getNotifiedKey(Data key) {
+        ConditionKey conditionKey = getSignalKey(key);
+        if (conditionKey == null) {
+            return new LockWaitNotifyKey(namespace, key);
+        }
+        return conditionKey;
     }
 
-    public void registerSignalKey(ConditionKey conditionKey) {
-        LockResourceImpl lock = getLock(conditionKey.getKey());
-        lock.registerSignalKey(conditionKey);
-    }
-
-    public ConditionKey getSignalKey(Data key) {
+    private ConditionKey getSignalKey(Data key) {
         LockResourceImpl lock = locks.get(key);
         if (lock == null) {
             return null;
@@ -297,6 +301,14 @@ public final class LockStoreImpl implements DataSerializable, LockStore {
         if (lock != null) {
             lock.removeSignalKey(conditionKey);
         }
+    }
+
+    public boolean hasSignalKey(ConditionKey conditionKey) {
+        LockResourceImpl lock = locks.get(conditionKey.getKey());
+        if (lock == null) {
+            return false;
+        }
+        return lock.hasSignalKey(conditionKey);
     }
 
     public void registerExpiredAwaitOp(AwaitOperation awaitResponse) {

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java
@@ -21,52 +21,41 @@ import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.DataSerializable;
 
 import java.io.IOException;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.HashSet;
+import java.util.Set;
 
-final class ConditionInfo implements DataSerializable {
+final class WaitersInfo implements DataSerializable {
 
     private String conditionId;
-    private Map<ConditionWaiter, ConditionWaiter> waiters = new HashMap<ConditionWaiter, ConditionWaiter>(2);
+    private Set<ConditionWaiter> waiters = new HashSet<ConditionWaiter>(2);
 
-    public ConditionInfo() {
+    public WaitersInfo() {
     }
 
-    public ConditionInfo(String conditionId) {
+    public WaitersInfo(String conditionId) {
         this.conditionId = conditionId;
     }
 
-    public boolean addWaiter(String caller, long threadId) {
+    public void addWaiter(String caller, long threadId) {
         ConditionWaiter waiter = new ConditionWaiter(caller, threadId);
-        return waiters.put(waiter, waiter) == null;
+        waiters.add(waiter);
     }
 
-    public boolean removeWaiter(String caller, long threadId) {
+    public void removeWaiter(String caller, long threadId) {
         ConditionWaiter waiter = new ConditionWaiter(caller, threadId);
-        return waiters.remove(waiter) != null;
+        waiters.remove(waiter);
     }
 
     public String getConditionId() {
         return conditionId;
     }
 
-    public int getAwaitCount() {
-        return waiters.size();
+    public boolean hasWaiter() {
+        return !waiters.isEmpty();
     }
 
-    public boolean startWaiter(String caller, long threadId) {
-        ConditionWaiter key = new ConditionWaiter(caller, threadId);
-        ConditionWaiter waiter = waiters.get(key);
-        if (waiter == null) {
-            throw new IllegalStateException();
-        }
-
-        if (waiter.started) {
-            return false;
-        } else {
-            waiter.started = true;
-            return true;
-        }
+    public Set<ConditionWaiter> getWaiters() {
+        return waiters;
     }
 
     @Override
@@ -75,7 +64,7 @@ final class ConditionInfo implements DataSerializable {
         int len = waiters.size();
         out.writeInt(len);
         if (len > 0) {
-            for (ConditionWaiter w : waiters.values()) {
+            for (ConditionWaiter w : waiters) {
                 out.writeUTF(w.caller);
                 out.writeLong(w.threadId);
             }
@@ -89,19 +78,26 @@ final class ConditionInfo implements DataSerializable {
         if (len > 0) {
             for (int i = 0; i < len; i++) {
                 ConditionWaiter waiter = new ConditionWaiter(in.readUTF(), in.readLong());
-                waiters.put(waiter, waiter);
+                waiters.add(waiter);
             }
         }
     }
 
-    private static class ConditionWaiter {
+    public static class ConditionWaiter {
         private final String caller;
         private final long threadId;
-        private transient boolean started;
 
         ConditionWaiter(String caller, long threadId) {
             this.caller = caller;
             this.threadId = threadId;
+        }
+
+        public long getThreadId() {
+            return threadId;
+        }
+
+        public String getCaller() {
+            return caller;
         }
 
         @Override

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AbstractLockOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AbstractLockOperation.java
@@ -104,12 +104,12 @@ public abstract class AbstractLockOperation extends AbstractOperation
         }
     }
 
-    protected final void setReferenceCallId(long refCallId) {
-        this.referenceCallId = refCallId;
-    }
-
     protected final long getReferenceCallId() {
         return referenceCallId != 0 ? referenceCallId : getCallId();
+    }
+
+    protected final void setReferenceCallId(long refCallId) {
+        this.referenceCallId = refCallId;
     }
 
     @Override
@@ -149,7 +149,7 @@ public abstract class AbstractLockOperation extends AbstractOperation
     @Override
     protected void toString(StringBuilder sb) {
         super.toString(sb);
-
         sb.append(", namespace=").append(namespace);
+        sb.append(", threadId=").append(threadId);
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AwaitBackupOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AwaitBackupOperation.java
@@ -47,7 +47,7 @@ public class AwaitBackupOperation extends AbstractLockOperation
     public void run() throws Exception {
         LockStoreImpl lockStore = getLockStore();
         lockStore.lock(key, originalCaller, threadId, getReferenceCallId(), leaseTime);
-        ConditionKey conditionKey = new ConditionKey(namespace.getObjectName(), key, conditionId);
+        ConditionKey conditionKey = new ConditionKey(namespace.getObjectName(), key, conditionId, originalCaller, threadId);
         lockStore.removeSignalKey(conditionKey);
         lockStore.removeAwait(key, conditionId, originalCaller, threadId);
         response = true;

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AwaitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AwaitOperation.java
@@ -44,14 +44,6 @@ public class AwaitOperation extends AbstractLockOperation
     }
 
     @Override
-    public void beforeRun() throws Exception {
-        if (!expired) {
-            LockStoreImpl lockStore = getLockStore();
-            lockStore.startAwaiting(key, conditionId, getCallerUuid(), threadId);
-        }
-    }
-
-    @Override
     public void run() throws Exception {
         LockStoreImpl lockStore = getLockStore();
         if (!lockStore.lock(key, getCallerUuid(), threadId, getReferenceCallId(), leaseTime)) {
@@ -70,25 +62,18 @@ public class AwaitOperation extends AbstractLockOperation
 
     @Override
     public ConditionKey getWaitKey() {
-        return new ConditionKey(namespace.getObjectName(), key, conditionId);
+        return new ConditionKey(namespace.getObjectName(), key, conditionId, getCallerUuid(), threadId);
     }
 
     @Override
     public boolean shouldWait() {
         LockStoreImpl lockStore = getLockStore();
-
-        ConditionKey signalKey = lockStore.getSignalKey(key);
-        if (signalKey == null) {
-            return true;
-        }
-
         boolean canAcquireLock = lockStore.canAcquireLock(key, getCallerUuid(), threadId);
-
         if (!canAcquireLock) {
             return true;
         }
 
-        return !conditionId.equals(signalKey.getConditionId());
+        return !lockStore.hasSignalKey(getWaitKey());
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/BaseSignalOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/BaseSignalOperation.java
@@ -16,7 +16,6 @@
 
 package com.hazelcast.concurrent.lock.operations;
 
-import com.hazelcast.concurrent.lock.ConditionKey;
 import com.hazelcast.concurrent.lock.LockStoreImpl;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
@@ -42,23 +41,11 @@ abstract class BaseSignalOperation extends AbstractLockOperation {
 
     @Override
     public void run() throws Exception {
-        LockStoreImpl lockStore = getLockStore();
-        awaitCount = lockStore.getAwaitCount(key, conditionId);
-        int signalCount;
-        if (awaitCount > 0) {
-            if (all) {
-                signalCount = awaitCount;
-            } else {
-                signalCount = 1;
-            }
-        } else {
-            signalCount = 0;
-        }
-        ConditionKey notifiedKey = new ConditionKey(namespace.getObjectName(), key, conditionId);
-        for (int i = 0; i < signalCount; i++) {
-            lockStore.registerSignalKey(notifiedKey);
-        }
         response = true;
+
+        LockStoreImpl lockStore = getLockStore();
+        int signalCount = all ? Integer.MAX_VALUE : 1;
+        lockStore.signal(key, conditionId, signalCount);
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/BeforeAwaitOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/BeforeAwaitOperation.java
@@ -18,7 +18,6 @@ package com.hazelcast.concurrent.lock.operations;
 
 import com.hazelcast.concurrent.lock.LockDataSerializerHook;
 import com.hazelcast.concurrent.lock.LockStoreImpl;
-import com.hazelcast.concurrent.lock.LockWaitNotifyKey;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.Data;
@@ -80,7 +79,8 @@ public class BeforeAwaitOperation extends AbstractLockOperation implements Notif
 
     @Override
     public WaitNotifyKey getNotifiedKey() {
-        return new LockWaitNotifyKey(namespace, key);
+        LockStoreImpl lockStore = getLockStore();
+        return lockStore.getNotifiedKey(key);
     }
 
     @Override

--- a/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/UnlockOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/UnlockOperation.java
@@ -16,10 +16,8 @@
 
 package com.hazelcast.concurrent.lock.operations;
 
-import com.hazelcast.concurrent.lock.ConditionKey;
 import com.hazelcast.concurrent.lock.LockDataSerializerHook;
 import com.hazelcast.concurrent.lock.LockStoreImpl;
-import com.hazelcast.concurrent.lock.LockWaitNotifyKey;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.Data;
@@ -114,12 +112,7 @@ public class UnlockOperation extends AbstractLockOperation implements Notifier, 
     @Override
     public final WaitNotifyKey getNotifiedKey() {
         LockStoreImpl lockStore = getLockStore();
-        ConditionKey conditionKey = lockStore.getSignalKey(key);
-        if (conditionKey == null) {
-            return new LockWaitNotifyKey(namespace, key);
-        } else {
-            return conditionKey;
-        }
+        return lockStore.getNotifiedKey(key);
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/concurrent/lock/ConditionAdvancedTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/concurrent/lock/ConditionAdvancedTest.java
@@ -161,7 +161,6 @@ public class ConditionAdvancedTest extends HazelcastTestSupport {
 
 
     @Test
-    @Ignore // https://github.com/hazelcast/hazelcast/issues/2272
     public void testLockConditionSignalAllShutDownKeyOwner() throws InterruptedException {
         final TestHazelcastInstanceFactory nodeFactory = createHazelcastInstanceFactory(2);
         final String name = randomString();

--- a/hazelcast/src/test/java/com/hazelcast/concurrent/lock/ProducerConsumerConditionStressTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/concurrent/lock/ProducerConsumerConditionStressTest.java
@@ -13,7 +13,6 @@ import static org.junit.Assert.assertNull;
 
 
 @RunWith(HazelcastParallelClassRunner.class)
-@Ignore // https://github.com/hazelcast/hazelcast/issues/2272
 public class ProducerConsumerConditionStressTest extends HazelcastTestSupport {
 
     private static volatile Object object;
@@ -27,20 +26,19 @@ public class ProducerConsumerConditionStressTest extends HazelcastTestSupport {
     public void test() {
         HazelcastInstance[] instances = createHazelcastInstanceFactory(INSTANCE_COUNT).newInstances();
         HazelcastInstance hz = instances[0];
-        //Hazelcast.newHazelcastInstance();
         ILock lock = hz.getLock(randomString());
         ICondition condition = lock.newCondition(randomString());
 
         ConsumerThread[] consumers = new ConsumerThread[CONSUMER_COUNT];
         for (int k = 0; k < consumers.length; k++) {
-            ConsumerThread thread = new ConsumerThread(1, lock, condition);
+            ConsumerThread thread = new ConsumerThread(k, lock, condition);
             thread.start();
             consumers[k] = thread;
         }
 
         ProducerThread[] producers = new ProducerThread[PRODUCER_COUNT];
         for (int k = 0; k < producers.length; k++) {
-            ProducerThread thread = new ProducerThread(1, lock, condition);
+            ProducerThread thread = new ProducerThread(k, lock, condition);
             thread.start();
             producers[k] = thread;
         }

